controllers/bucket_controller_test.go

```diff
@@ -17,59 +17,632 @@ limitations under the License.
+			},
+		},
+		{
+			name: "updates hostname on diff from current",
+			beforeFunc: func(obj *sourcev1.Bucket, storage *Storage) error {
+				obj.Status.Artifact = &sourcev1.Artifact{
+					Path:     fmt.Sprintf("/reconcile-storage/hostname.txt"),
+					Revision: "f",
+					Checksum: "3b9c358f36f0a31b6ad3e14f309c7cf198ac9246e8316f9ce543d5b19ac02b80",
+					URL:      "http://outdated.com/reconcile-storage/hostname.txt",
+				}
+				if err := testStorage.MkdirAll(*obj.Status.Artifact); err != nil {
+					return err
+				}
+				if err := testStorage.AtomicWriteFile(obj.Status.Artifact, strings.NewReader("file"), 0644); err != nil {
+					return err
+				}
+				return nil
+			},
+			assertPaths: []string{
+				"/reconcile-storage/hostname.txt",
+			},
+			assertArtifact: &sourcev1.Artifact{
+				Path:     "/reconcile-storage/hostname.txt",
+				Revision: "f",
+				Checksum: "3b9c358f36f0a31b6ad3e14f309c7cf198ac9246e8316f9ce543d5b19ac02b80",
+				URL:      testStorage.Hostname + "/reconcile-storage/hostname.txt",
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			r := &BucketReconciler{
+				Storage: testStorage,
+			}
+
+			obj := &sourcev1.Bucket{
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: "test-",
+				},
+			}
+			if tt.beforeFunc != nil {
+				g.Expect(tt.beforeFunc(obj, testStorage)).To(Succeed())
+			}
+
+			got, err := r.reconcileStorage(context.TODO(), obj)
+			g.Expect(err != nil).To(Equal(tt.wantErr))
+			g.Expect(got).To(Equal(tt.want))
+
+			g.Expect(obj.Status.Artifact).To(MatchArtifact(tt.assertArtifact))
+			if tt.assertArtifact != nil && tt.assertArtifact.URL != "" {
+				g.Expect(obj.Status.Artifact.URL).To(Equal(tt.assertArtifact.URL))
+			}
+			g.Expect(obj.Status.Conditions).To(conditions.MatchConditions(tt.assertConditions))
+
+			for _, p := range tt.assertPaths {
+				absoluteP := filepath.Join(testStorage.BasePath, p)
+				if !strings.HasPrefix(p, "!") {
+					g.Expect(absoluteP).To(BeAnExistingFile())
+					continue
+				}
+				g.Expect(absoluteP).NotTo(BeAnExistingFile())
+			}
+		})
+	}
+}
+
+func TestBucketReconciler_reconcileSource(t *testing.T) {
+	tests := []struct {
+		name             string
+		bucketName       string
+		bucketObjects    []*s3MockObject
+		middleware       http.Handler
+		secret           *corev1.Secret
+		beforeFunc       func(obj *sourcev1.Bucket)
+		want             ctrl.Result
+		wantErr          bool
+		assertArtifact   sourcev1.Artifact
+		assertConditions []metav1.Condition
+	}{
+		{
+			name:       "reconciles source",
+			bucketName: "dummy",
+			bucketObjects: []*s3MockObject{
+				{
+					Key:          "test.txt",
+					Content:      []byte("test"),
+					ContentType:  "text/plain",
+					LastModified: time.Now(),
+				},
+			},
+			assertArtifact: sourcev1.Artifact{
+				Path:     "bucket/test-bucket/f0467900d3cede8323f3e61a1467f7cd370d1c0d942ff990a1a7be1eb1a231e8.tar.gz",
```

Comment:
Buckets are namespaced objects so this path is wrong, we should have tests that make sure this can't happen.

Reply:
Based on further discussions in private the storage logic should scream early on when an object does not contain a namespace, requiring the caller to either provide a mock with a namespace set, or let it pass the Kubernetes API first to ensure defaults are set.

Reply:
I have attempted to do this but the changes still touch too many files that are going to be changed in the future to not break the old tests, will re-attempt in a separate PR later on.